libs/native/src/main/java/org/elasticsearch/nativeaccess/PosixNativeAccess.java

```diff
@@ -121,6 +125,49 @@ public void tryLockMemory() {
 
     protected abstract void logMemoryLimitInstructions();
 
+    @Override
+    public OptionalLong allocatedSizeInBytes(Path path) {
+        assert Files.isRegularFile(path) : path;
+        var stats = libc.newStat64(constants.statStructSize(), constants.statStructSizeOffset(), constants.statStructBlocksOffset());
+
+        int fd = libc.open(path.toAbsolutePath().toString(), O_WRONLY, constants.O_CREAT());
+        if (fd == -1) {
+            logger.warn("Could not open file [" + path + "] to get allocated size: " + libc.strerror(libc.errno()));
+            return OptionalLong.empty();
+        }
+
+        if (libc.fstat64(fd, stats) != 0) {
```

Comment:
This is failing in the tests; in the original code, xstat was used.
I don't think this is relevant - maybe xstat can get a file size for a file that was just pre-allocated, and fstat cannot, but I seriously doubt it.
The errno is `Invalid Argument`. I'd double check the offsets for the struct here.

Reply:
This was a bug in how the file was opened. I copy/pasted the open code and used O_WRONLY, when it should have been O_RDONLY.

Reply:
Hrm, or at least that's what I thought. I got bit by an assume that was making this test not run locally. I'll keep digging.

Reply:
I see you switched to Memory; good call. It seems that even if you use a JNA `Structure` byte[] are not inlined in the struct, so the padding was not right. Better use a block of memory and access the correct offset.